File: client/src/game/components/wave.lua
local constants = require 'src.game.components.constants'

local Wave = {
  _meta = constants.Wave
}

local types = {
  CIRCULAR = 'circular',
  HORIZONTAL = 'horizontal',
  VERTICAL = 'vertical'
}

function Wave.new(id, type, x, y, amplitude, frequency)
  local wave = {
    _meta = Wave._meta,
    id = id,
    type = type,
    x = x or 0,
    y = y or 0,
    amplitude = amplitude or 1,
    frequency = frequency or 1
  }

  return wave
end

function Wave.newCircular(id, x, y, amplitude, frequency)
  local wave = Wave.new(id, types.CIRCULAR, x, y, amplitude, frequency)

  return wave
end

function Wave.newHorizontal(id, x, y, amplitude, frequency, direction)
  local wave = Wave.new(id, types.HORIZONTAL, x, y, amplitude, frequency)
  wave.direction = direction or 1

  return wave
end

function Wave.newVertical(id, x, y, amplitude, frequency)
  local wave = Wave.new(id, types.VERTICAL, x, y, amplitude, frequency)
  wave.direction = direction or 1

  return wave
end

return Wave

File: client2/src/game/components/wave.ts
import { Flag } from './flags'

enum WaveType {
  Circular = 'circular',
  Horizontal = 'horizontal',
  Vertical = 'vertical',
}

export class Wave {
  static _id = 'Wave'
  _id = Wave._id

  static _flag = Flag.Wave
  _flag = Flag.Wave

  type: WaveType
  x: number
  y: number
  amplitude: number
  frequency: number
  direction: number

  constructor(
    type: WaveType = WaveType.Vertical,
    x: number = 0,
    y: number = 0,
    amplitude: number = 0,
    frequency: number = 0,
    direction: number = 0
  ) {
    this.type = type
    this.x = x
    this.y = y
    this.amplitude = amplitude
    this.frequency = frequency
    this.direction = direction
  }
}

export class CircularWave extends Wave {
  constructor(
    x: number = 0,
    y: number = 0,
    amplitude: number = 0,
    frequency: number = 0
  ) {
    super(WaveType.Circular, x, y, amplitude, frequency)
  }
}

export class HorizontalWave extends Wave {
  constructor(
    x: number = 0,
    y: number = 0,
    amplitude: number = 0,
    frequency: number = 0,
    direction: number = 1
  ) {
    super(WaveType.Horizontal, x, y, amplitude, frequency, direction)
  }
}

export class VerticalWave extends Wave {
  constructor(
    x: number = 0,
    y: number = 0,
    amplitude: number = 0,
    frequency: number = 0,
    direction: number = 1
  ) {
    super(WaveType.Vertical, x, y, amplitude, frequency, direction)
  }
}
